scripts/provision-group-check-users.js

```diff
@@ -0,0 +1,319 @@
+  }
+
+  return {usersByEmail, usersByUsername};
+}
+
+
+/**
+ * For a given nextstrain group and username, return the list of roles this username
+ * already has for the (nextstrain) group. In theory a user should only have a single
+ * role, but there's nothing enforcing this as far as I'm aware, so we return an array
+ */
+async function getUserRoles(group, username) {
+  // roles are based upon group membership, which must be queried per-username
+  const params = {
+    Username: username,
+    UserPoolId: COGNITO_USER_POOL_ID,
+  }
+  let cognitoGroups = [];
+  try {
+    const res = await cognito.send(new AdminListGroupsForUserCommand(params));
+    cognitoGroups = res.Groups.map((g) => g.GroupName);
+    if (!Array.isArray(cognitoGroups)) {
+      throw new Error(`unexpected type for "Groups" data from the AdminListGroupsForUserCommand request for user ${username}`)
+    }
+  } catch (e) {
+    console.error(`FATAL ERROR while fetching (cognito) groups for user ${username}`)
+    console.error(e)
+    process.exit(2)
+  }
+
+  // for the specified nextstrain group, which cognito groups are relevant?
+  // (each cognito group corresponds to a role)
+  const rolesByCognitoGroup = new Map(
+    Array.from(group.membershipRoles.entries())
+      .map(([role, groupName]) => [groupName, role])
+  );
+
+  const existingRoles = cognitoGroups
+    .map((g) => rolesByCognitoGroup.has(g) ? rolesByCognitoGroup.get(g) : false)
+    .filter(Boolean);
+
+  return existingRoles;
+}
+
+/**
+ * The cognito pools we use are case-sensitive for emails, but we want to avoid
+ * provisioning multiple emails with different casings. Here we create a lookup
+ * dictionary by uppercase email addresses, and for good measure spit out warnings
+ * if there are multiple emails with different casings.
+ */
+function summariseEmails(usersByEmail) {
+  const db = {}
+  for (const email of Object.keys(usersByEmail)) {
+    const key = email.toUpperCase();
+    Object.hasOwn(db, key) ? db[key].push(email) : (db[key]=[email]);
+  }
+  // eslint-disable-next-line no-unused-vars
+  for (const [_key, emails] of Object.entries(db)) {
+    if (emails.length>1) {
+      console.log(`WARNING: Multiple emails with different casings present in pool: ${emails.join(', ')}`)
+    }
+  }
+  return db;
+}
+
+/**
+ * Match a provided case-sensitive *email* against the *emailDb* which is
+ * indexed by all-caps emails. This is necessary because cognito (at least the
+ * way we have it) is case-sensitive, but emails are generally case-insensitive.
+ * Ultimately we want to avoid situations where we provision multiple emails
+ * which differ only in their casing.
+ * @param {object} emailDb
+ * @param {string} email
+ * @returns {object} exact: boolean (whether the email is an exact match),
+ *  candidates: string[] cognito URLs which are identical to the provided email
+ *  except for potential different casing
+ */
+function searchEmails(emailDb, email) {
```

Comment:
If the warning that's given as a byproduct of creating `emailDb` is not needed (https://github.com/nextstrain/nextstrain.org/pull/987#discussion_r2520458024), the code could be simplified to remove the entire `emailDb` intermediate data structure and instead do case-insensitive searches on-the-fly. Example: a1a599f8c5ad8cf87258e086ef39d94405d73545

Reply:
(non-blocking) Another follow-up suggested simplification: 9332153280e419dfe7c512d8843e984eaf8ca656